README.md
# Repository Coverage

[Full report](https://htmlpreview.github.io/?https://github.com/curveresearch/curvesim/blob/python-coverage-comment-action-data/htmlcov/index.html)

| Name                                                                |    Stmts |     Miss |   Branch |   BrPart |   Cover |   Missing |
|-------------------------------------------------------------------- | -------: | -------: | -------: | -------: | ------: | --------: |
| curvesim/\_\_init\_\_.py                                            |        5 |        0 |        0 |        0 |    100% |           |
| curvesim/\_order\_book/\_\_init\_\_.py                              |       53 |       48 |       12 |        0 |      8% |53-101, 105-131 |
| curvesim/constants.py                                               |       15 |        0 |        0 |        0 |    100% |           |
| curvesim/exceptions/\_\_init\_\_.py                                 |       23 |        0 |        0 |        0 |    100% |           |
| curvesim/iterators/\_\_init\_\_.py                                  |        0 |        0 |        0 |        0 |    100% |           |
| curvesim/iterators/param\_samplers/\_\_init\_\_.py                  |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/iterators/param\_samplers/parameterized\_pool\_iterator.py |       51 |        1 |       10 |        1 |     97% |        40 |
| curvesim/iterators/param\_samplers/pool\_mixins.py                  |       51 |        0 |       16 |        6 |     91% |23->22, 27->26, 44->43, 48->47, 65->64, 69->68 |
| curvesim/iterators/price\_samplers/\_\_init\_\_.py                  |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/iterators/price\_samplers/price\_volume.py                 |       23 |        0 |        6 |        1 |     97% |    73->72 |
| curvesim/logging.py                                                 |       47 |        1 |       10 |        3 |     93% |31, 96->98, 104->103 |
| curvesim/metrics/\_\_init\_\_.py                                    |        6 |        0 |        2 |        0 |    100% |           |
| curvesim/metrics/base.py                                            |      128 |        6 |       69 |       19 |     87% |95->94, 108->107, 129->132, 130->129, 131->130, 132->131, 152->154, 153->152, 154->153, 170->172, 171->170, 172->171, 221, 279->282, 280->279, 281->280, 282->281, 291-294, 379->exit, 380 |
| curvesim/metrics/metrics.py                                         |      226 |        0 |       58 |       14 |     95% |39->38, 155->157, 156->155, 157->156, 267->269, 268->267, 269->268, 336->338, 337->336, 338->337, 525->527, 526->525, 527->526, 655->654 |
| curvesim/metrics/results/\_\_init\_\_.py                            |        3 |        0 |        0 |        0 |    100% |           |
| curvesim/metrics/results/make\_results.py                           |       28 |        0 |       10 |        0 |    100% |           |
| curvesim/metrics/results/sim\_results.py                            |       35 |       12 |       12 |        3 |     55% |63-66, 92-94, 121, 129-132 |
| curvesim/metrics/state\_log/\_\_init\_\_.py                         |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/metrics/state\_log/log.py                                  |       32 |        0 |       14 |        2 |     96% |35->34, 54->53 |
| curvesim/metrics/state\_log/pool\_parameters.py                     |       28 |        0 |       10 |        0 |    100% |           |
| curvesim/metrics/state\_log/pool\_state.py                          |       18 |        0 |        2 |        0 |    100% |           |
| curvesim/network/\_\_init\_\_.py                                    |        0 |        0 |        0 |        0 |    100% |           |
| curvesim/network/coingecko.py                                       |       80 |        6 |       12 |        2 |     91% |63-67, 153 |
| curvesim/network/curve\_prices.py                                   |       34 |        3 |        4 |        2 |     87% |47-48, 118->121, 122 |
| curvesim/network/http.py                                            |       39 |        5 |       20 |        7 |     80% |15->17, 16->15, 17->16, 37->39, 38->37, 39->38, 42->45, 49-54 |
| curvesim/network/subgraph.py                                        |      124 |       48 |       48 |        6 |     59% |59-62, 94, 129-160, 321, 323, 325, 328->331, 385-411, 444-468 |
| curvesim/network/utils.py                                           |       22 |        6 |        6 |        2 |     64% |21-22, 41->40, 48-51 |
| curvesim/network/web3.py                                            |       70 |        5 |       16 |        5 |     86% |62-63, 139->147, 143-144, 147->152, 176 |
| curvesim/overrides/\_\_init\_\_.py                                  |        9 |        3 |        4 |        1 |     54% |     40-42 |
| curvesim/pipelines/\_\_init\_\_.py                                  |       21 |        5 |       12 |        1 |     76% |     75-79 |
| curvesim/pipelines/common/\_\_init\_\_.py                           |       42 |        4 |       10 |        0 |     92% |     78-89 |
| curvesim/pipelines/simple/\_\_init\_\_.py                           |       20 |        0 |        0 |        0 |    100% |           |
| curvesim/pipelines/simple/strategy.py                               |       11 |        0 |        0 |        0 |    100% |           |
| curvesim/pipelines/simple/trader.py                                 |       26 |        0 |       10 |        0 |    100% |           |
| curvesim/pipelines/vol\_limited\_arb/\_\_init\_\_.py                |       29 |       17 |        4 |        0 |     36% |    90-115 |
| curvesim/pipelines/vol\_limited\_arb/strategy.py                    |       22 |       10 |        6 |        0 |     43% |30-31, 34-35, 39-46 |
| curvesim/pipelines/vol\_limited\_arb/trader.py                      |       81 |       57 |       34 |        0 |     24% |41-44, 76-133, 159-160, 168-174, 204-216 |
| curvesim/plot/\_\_init\_\_.py                                       |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/plot/altair/\_\_init\_\_.py                                |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/plot/altair/chart\_properties.py                           |       35 |       25 |       14 |        0 |     20% |31-38, 61-66, 83-88, 130-142 |
| curvesim/plot/altair/make\_chart.py                                 |       30 |       23 |       10 |        0 |     18% |48-62, 94-117 |
| curvesim/plot/altair/results/\_\_init\_\_.py                        |        0 |        0 |        0 |        0 |    100% |           |
| curvesim/plot/altair/results/make\_page.py                          |       60 |       46 |       18 |        0 |     18% |40-46, 71-73, 98-109, 128-142, 167-172, 197-200, 225-237 |
| curvesim/plot/altair/results/preprocessing.py                       |       40 |       33 |       12 |        0 |     13% |32-38, 64-83, 104-106, 125-128, 147-154 |
| curvesim/plot/altair/results/result\_plotter.py                     |       25 |       14 |        2 |        1 |     44% |15-17, 25-30, 38-43, 52->51, 53 |
| curvesim/plot/altair/results/result\_selectors.py                   |       40 |       32 |        6 |        0 |     17% |31-43, 65-75, 94-105, 127-137, 159-169, 188-190 |
| curvesim/plot/altair/results/tooltip.py                             |       11 |        9 |        4 |        0 |     13% |      6-18 |
| curvesim/plot/altair/selectors.py                                   |       32 |       24 |       10 |        0 |     19% |59-69, 97-109, 134-154, 179 |
| curvesim/plot/altair/styles.py                                      |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/plot/result\_plotter.py                                    |       16 |        9 |        8 |        0 |     29% |     53-63 |
| curvesim/pool/\_\_init\_\_.py                                       |       67 |       13 |       28 |        9 |     73% |111, 180, 184-185, 188, 259, 263-264, 267, 278-281, 297->306 |
| curvesim/pool/base.py                                               |       51 |       21 |       34 |       12 |     49% |29->28, 31-33, 36->35, 38-40, 43->42, 47, 50->49, 54, 57->56, 59-61, 64->63, 68, 71->70, 73-75, 78->77, 82, 87-91 |
| curvesim/pool/cryptoswap/\_\_init\_\_.py                            |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/pool/cryptoswap/calcs/\_\_init\_\_.py                      |       68 |        6 |       28 |        6 |     88% |39, 58, 76, 100, 137, 141 |
| curvesim/pool/cryptoswap/calcs/factory\_2\_coin.py                  |      126 |       13 |       46 |       12 |     85% |40, 60, 102, 104, 142-143, 156, 166, 185, 187, 230, 239, 242 |
| curvesim/pool/cryptoswap/calcs/tricrypto\_ng.py                     |      311 |       28 |      122 |       29 |     87% |68->72, 114, 116, 118, 120, 134, 136, 150, 152, 157, 163, 167, 173, 181, 187, 239, 241, 285-286, 299, 309, 321, 441, 445, 523, 534, 541, 578, 605->608, 630 |
| curvesim/pool/cryptoswap/pool.py                                    |      475 |       32 |      160 |       21 |     91% |169, 176, 179, 184->187, 190, 204, 341, 394->403, 399, 474->481, 650->663, 654, 710, 757, 776-791, 803-807, 862->865, 999-1001, 1019->1029, 1083, 1108 |
| curvesim/pool/sim\_interface/\_\_init\_\_.py                        |        5 |        0 |        0 |        0 |    100% |           |
| curvesim/pool/sim\_interface/asset\_indices.py                      |       22 |        1 |       16 |        5 |     84% |32->31, 36->38, 37->36, 38->37, 60 |
| curvesim/pool/sim\_interface/cryptoswap.py                          |       83 |        1 |       50 |       19 |     85% |29, 33->36, 34->33, 35->34, 36->35, 40->42, 41->40, 42->41, 47->46, 77->76, 108->107, 139->138, 160->159, 174->173, 191->195, 217->220, 218->217, 219->218, 220->219 |
| curvesim/pool/sim\_interface/metapool.py                            |       90 |        3 |       54 |       18 |     85% |30, 32->35, 33->32, 34->33, 35->34, 47->49, 48->47, 49->48, 58->57, 96->95, 144, 147, 154->153, 202->201, 218->221, 219->218, 220->219, 221->220 |
| curvesim/pool/sim\_interface/pool.py                                |       46 |        1 |       34 |       16 |     79% |22, 24->27, 25->24, 26->25, 27->26, 31->33, 32->31, 33->32, 38->37, 69->68, 100->99, 129->128, 145->148, 146->145, 147->146, 148->147 |
| curvesim/pool/sim\_interface/raipool.py                             |       10 |        2 |        0 |        0 |     80% |     34-35 |
| curvesim/pool/snapshot.py                                           |       89 |        0 |       10 |        4 |     96% |76->75, 107->106, 139->138, 191->190 |
| curvesim/pool/stableswap/\_\_init\_\_.py                            |        4 |        0 |        0 |        0 |    100% |           |
| curvesim/pool/stableswap/metapool.py                                |      291 |       11 |       90 |        9 |     95% |88, 99, 371, 507->519, 526, 580->579, 677-679, 711, 797, 818, 875 |
| curvesim/pool/stableswap/pool.py                                    |      246 |        4 |       66 |        4 |     97% |84, 410->422, 429, 554, 669 |
| curvesim/pool/stableswap/raipool.py                                 |       14 |        8 |        2 |        0 |     38% |35-41, 44-46 |
| curvesim/pool\_data/\_\_init\_\_.py                                 |        3 |        0 |        0 |        0 |    100% |           |
| curvesim/pool\_data/metadata/\_\_init\_\_.py                        |       35 |        5 |        6 |        1 |     85% |55, 66-67, 72-73 |
| curvesim/pool\_data/metadata/base.py                                |       26 |        0 |        8 |        4 |     88% |152->151, 156->155, 160->159, 164->163 |
| curvesim/pool\_data/metadata/cryptoswap.py                          |       25 |        0 |       12 |        3 |     92% |46->45, 50->49, 54->53 |
| curvesim/pool\_data/metadata/stableswap.py                          |       43 |        0 |       20 |        3 |     95% |51->50, 62->61, 73->72 |
| curvesim/pool\_data/queries/\_\_init\_\_.py                         |        0 |        0 |        0 |        0 |    100% |           |
| curvesim/pool\_data/queries/metadata.py                             |       26 |        3 |        4 |        0 |     90% |     54-58 |
| curvesim/pool\_data/queries/pool\_volume.py                         |       56 |        1 |       10 |        2 |     95% |93, 124->126 |
| curvesim/price\_data/\_\_init\_\_.py                                |       10 |        4 |        6 |        1 |     44% |     59-63 |
| curvesim/price\_data/sources.py                                     |        8 |        0 |        0 |        0 |    100% |           |
| curvesim/sim/\_\_init\_\_.py                                        |       29 |       22 |       13 |        0 |     17% |140-152, 156-193 |
| curvesim/templates/\_\_init\_\_.py                                  |        8 |        0 |        0 |        0 |    100% |           |
| curvesim/templates/log.py                                           |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/templates/param\_samplers.py                               |       36 |        0 |       18 |        1 |     98% |    56->55 |
| curvesim/templates/price\_samplers.py                               |       10 |        0 |        2 |        0 |    100% |           |
| curvesim/templates/sim\_assets.py                                   |        8 |        0 |        0 |        0 |    100% |           |
| curvesim/templates/sim\_pool.py                                     |        8 |        0 |        0 |        0 |    100% |           |
| curvesim/templates/strategy.py                                      |       23 |        0 |        2 |        0 |    100% |           |
| curvesim/templates/trader.py                                        |       47 |        3 |       16 |        1 |     87% |46, 50-51, 54->53 |
| curvesim/tools/\_\_init\_\_.py                                      |        2 |        0 |        0 |        0 |    100% |           |
| curvesim/tools/bonding\_curve.py                                    |       76 |       19 |       28 |        5 |     69% |70->91, 79->91, 89, 119-123, 145->149, 161-177 |
| curvesim/utils/\_\_init\_\_.py                                      |       37 |        3 |        8 |        3 |     87% |68, 71, 107 |
| curvesim/utils/address.py                                           |       12 |        3 |        2 |        0 |     64% |     47-50 |
| curvesim/utils/decorators.py                                        |       25 |        8 |       13 |        4 |     63% |41, 53-62, 64->exit, 65 |
| curvesim/version.py                                                 |        7 |        0 |        0 |        0 |    100% |           |
|                                                           **TOTAL** | **4265** |  **667** | **1409** |  **268** | **80%** |           |


## Setup coverage badge

Below are examples of the badges you can use in your main branch `README` file.

### Direct image

[![Coverage badge](https://raw.githubusercontent.com/curveresearch/curvesim/python-coverage-comment-action-data/badge.svg)](https://htmlpreview.github.io/?https://github.com/curveresearch/curvesim/blob/python-coverage-comment-action-data/htmlcov/index.html)

This is the one to use if your repository is private or if you don't want to customize anything.

### [Shields.io](https://shields.io) Json Endpoint

[![Coverage badge](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/curveresearch/curvesim/python-coverage-comment-action-data/endpoint.json)](https://htmlpreview.github.io/?https://github.com/curveresearch/curvesim/blob/python-coverage-comment-action-data/htmlcov/index.html)

Using this one will allow you to [customize](https://shields.io/endpoint) the look of your badge.
It won't work with private repositories. It won't be refreshed more than once per five minutes.

### [Shields.io](https://shields.io) Dynamic Badge

[![Coverage badge](https://img.shields.io/badge/dynamic/json?color=brightgreen&label=coverage&query=%24.message&url=https%3A%2F%2Fraw.githubusercontent.com%2Fcurveresearch%2Fcurvesim%2Fpython-coverage-comment-action-data%2Fendpoint.json)](https://htmlpreview.github.io/?https://github.com/curveresearch/curvesim/blob/python-coverage-comment-action-data/htmlcov/index.html)

This one will always be the same color. It won't work for private repos. I'm not even sure why we included it.

## What is that?

This branch is part of the
[python-coverage-comment-action](https://github.com/marketplace/actions/python-coverage-comment)
GitHub Action. All the files in this branch are automatically generated and may be
overwritten at any moment.

CHANGELOG.rst

.. _changelog-0.5.0.rc0:

0.5.0.rc0 — 2023-11-06
======================

Removed
-------

- Removed modes for volume multiplier; changes from the default multipliers
  must be specified precisely through arguments.
- Volume query is now done through the new `curve-prices` API instead of the
  Curve subgraph.  This means currently only Mainnet is supported for this query. 
- RAI3CRV pool is currently unsupported by simulation pipelines. It will
  be reimplemented along with Stableswap-NG pools.
- `PoolDataCache` has been removed.


Added
-----

- Use Curve Prices API for the volume query (Mainnet only).
- Added `PriceDepth` (liquidity density) metric for Cryptoswap pools.

- Fully enabled 3-coin cryptoswap functionality based on Tricrypto-NG code:

  - NG oracle behavior
  - `lp_price`
  - `get_dy`
  - `calc_withdraw_one_coin`
  - `add_liquidity`
  - `remove_liquidity` (now returns amounts withdrawn)
  - `remove_liquidity_one_coin`

- Stableswap pools now have:

  - `remove_liquidity`
  - `remove_liquidity_imbalance`

- `curvesim.bonding_curve` now handles Cryptoswap pools.


Changed
-------

- Improved liquidity density calculation for robustness across pool types

- Enabled `CurveCryptoPool._claim_admin_fees`. For maintainability, Tricrypto_ng's
  implementation and usage patterns, which are in `test/fixtures/curve/tricrypto_ng.vy`,
  are used for both 2-coin and 3-coin Cryptoswap pools.

- Volume multipliers now computed individually for each asset pair

- Arbitrage optimizers exclude trades smaller than a pool's minimum trade size
  from optimization. This reduces variability in optimization and prevents some
  potential errors.

- Price error changes:
  - normalized by target price
  - now a dict mapping trading pair to each error value


Fixed
-----

- Cryptoswap state snapshotting was updated to include additional price and profit-related
  attributes.  This is necessary for proper repegging mechanics during simulations.
- Fixed handling of integer signed division in Tricrypto-NG's `get_y` calc.



.. _changelog-0.5.0.b2:

0.5.0.b2 — 2023-09-12
=====================

Removed
-------

- Pool metadata no longer takes balancing options.  This is handled by the
  the appropriate factory function, namely `get_sim_pool`.

- Removed default and test parameters from simulation pipelines (#222, #247).
  Users will have to explicitly choose the pool parameters to vary for sims.


Added
-----

- `virtual_price` is an allowed optional argument for pools.

- Added fixtures for TriCrypto-NG factory and views contracts for exact testing of view functionality.
  These contracts are copyrighted by Swiss Stake GmbH and used with permission.

- Each pool now has an internal method to handle conversion of the invariant
  to balances.  This is helpful for use by "friend" classes/functions such
  as factory functions.


Changed
-------

- CurveCryptoPool can use calc_token_amount to determine LP tokens minted on deposits for 3-coin pools now.

- Reset cryptoswap profit counters to initial values at the start of each
  sim run (`SimCurveCryptoPool.prepare_for_run`). (#246)

- The pool snapshot is no longer augmented with calculated values and solely handles
  conversion of pulled subgraph data into appropriate python types.

  The needed calculations are done in the factory function `get_sim_pool` where they can
  be done rigorously by invoking the pool's internal logic.  This affects the rebalancing
  logic when initializing a pool for a sim run.

  The computation of `D` in the snapshot actually had a bug, so removing this fixes that.

- Test parameters are now defined within each pipeline test  (#247)

- Since `asyncio.get_event_loop` is deprecated in later versions of Python, we have replaced it by
  a custom function `curvesim.utils.get_event_loop`.


Fixed
-----

- The `CurveCryptoPool.__init__` had several bugs:
  - pool creation failed if no `tokens` argument provided (#219)
  - wrong `D` was computed for 3-coin pools when not explicitly passed-in (#219)
  - different price-related state variables were initialized with the same object (#237)

- Pool snapshot from subgraph was being augmented with some calculated values,
  namely `tokens` and `D`.
  These were incorrect for stableswap pools and metapools. (#239)

- Fixed bug in `SimCurveCryptoPool.prepare_for_run` where the different price-related
  state variables were initialized pointing to the same object. (#238)


.. _changelog-0.5.0.b1:

0.5.0.b1 — 2023-08-17
=====================

The cryptosim milestone enables simulation capability for cryptopools, such as 2-coin factory and 3-coin tricrypto-ng pools.

Note this is a **beta release** as some limitations exist due to data provider issues, which should resolve in the near future; in addition, while simulations are fully functioning, complete pythonic interaction with 3-coin cryptopools is not yet available.

Added
-----

- The `CurveCryptoPool` class was updated to handle 3 coins for the functionality required
  for cryptoswap simulations, namely `exchange`, `_tweak_price`, and the related calculations.
  The pythonic equivalents such as `add_liquidity`, `remove_liquidity_one_coin`, etc., were not updated.
  For supported calculations, the pool will give exact integer results as compared to the
  Tricrypto-NG contract.

- Add volume limiting in multiple denominations. Previously, volume
  limiting was only in USD. This update allows for volume limiting
  across multiple asset types (e.g., in v2 pools).

- Added `get_y` to `CurveCryptoPool` as with the stableswap pools.
  This breaks tight adherence to the vyper interface, but makes it easier
  for integrators.

- Added spot price methods `dydx` and `dydxfee` to the `CurveCryptoPool`.

- Cryptopools are now usable in the simulation framework, with some caveats.  Currently the Curve
  subgraph returns erroneous total volumes, but a workaround is to use the `vol_mult` argument to

nomics.py
import os
from datetime import timedelta, timezone
from itertools import combinations
from time import sleep

import pandas as pd
import requests
from dotenv import load_dotenv

load_dotenv()
key = os.environ.get("NOMICS_API_KEY")


def get_mkt(market, t_start, t_end):
    # Round times to nearest 30-min interval (floor)
    t_start = t_start.replace(minute=t_start.minute // 30 * 30, second=0, microsecond=0)
    t_end = t_end.replace(minute=t_end.minute // 30 * 30, second=0, microsecond=0)

    # Divide timerange into 14-day chunks
    t_starts = pd.date_range(start=t_start, end=t_end, freq="14d")
    t_ends = t_starts + timedelta(days=14) - timedelta(minutes=30)

    t_starts = t_starts.to_list()
    t_ends = t_ends.to_list()
    t_ends[-1] = t_end  # replace last value with end of total timerange

    # Request parameters
    url = "https://api.nomics.com/v1/exchange_candles"
    p = {"key": key, "interval": "30m", "exchange": market[0], "market": market[1]}

    # Requests
    data = []  # list of dataframes we concat at the end
    for i in range(len(t_starts)):
        p["start"] = t_starts[i].strftime("%Y-%m-%dT%H:%M:%SZ")
        p["end"] = t_ends[i].strftime("%Y-%m-%dT%H:%M:%SZ")

        getData = True
        while getData:
            r = requests.get(url, params=p)
            if r.status_code == 200:
                getData = False
            else:
                print(str(r.status_code))
                sleep(0.1)
        if r.json():
            data.append(pd.DataFrame(r.json())[["timestamp", "close", "volume"]].set_index("timestamp"))

    # Combine & Format data
    if data:
        data = pd.concat(data)
        data.columns = ["price", "volume"]
    else:
        data = pd.DataFrame(columns=["price", "volume"])
    data.index = pd.to_datetime(data.index)
    data["price"] = pd.to_numeric(data["price"]) ** market[2]  # take reciprocal when base/quote reversed
    if market[2] == 1:
        data["volume"] = pd.to_numeric(data["volume"])
    elif market[2] == -1:
        data["volume"] = pd.to_numeric(data["volume"]) / data["price"]

    # Fill in missing data with zeros
    t_samples = pd.date_range(start=t_start, end=t_end, freq="30min", tz=timezone.utc)
    data = data.reindex(t_samples, fill_value=0)

    return data


def get_agg(coins, t_start, t_end, exp=1):
    # Round times to nearest 5-min interval (floor)
    t_start = t_start.replace(minute=t_start.minute // 30 * 30, second=0, microsecond=0)
    t_end = t_end.replace(minute=t_end.minute // 30 * 30, second=0, microsecond=0)

    # Divide timerange into 3-day chunks
    t_starts = pd.date_range(start=t_start, end=t_end, freq="14d")
    t_ends = t_starts + timedelta(days=14) - timedelta(minutes=30)

    t_starts = t_starts.to_list()
    t_ends = t_ends.to_list()
    t_ends[-1] = t_end  # replace last value with end of total timerange

    # Request parameters
    url = "https://api.nomics.com/v1/markets/candles"
    p = {"key": key, "interval": "30m", "base": coins[0], "quote": coins[1]}

    # Requests
    data = []  # list of dataframes we concat at the end
    for i in range(len(t_starts)):
        p["start"] = t_starts[i].strftime("%Y-%m-%dT%H:%M:%SZ")
        p["end"] = t_ends[i].strftime("%Y-%m-%dT%H:%M:%SZ")

        getData = True
        while getData:
            r = requests.get(url, params=p)
            if r.status_code == 200:
                getData = False
            else:
                print(str(r.status_code))
                sleep(0.1)
        if r.json():
            data.append(pd.DataFrame(r.json())[["timestamp", "close", "volume"]].set_index("timestamp"))

    # Combine & Format data
    if data:
        data = pd.concat(data)
        data.columns = ["price", "volume"]
    else:
        data = pd.DataFrame(columns=["price", "volume"])

    data.index = pd.to_datetime(data.index)
    data["price"] = pd.to_numeric(data["price"]) ** exp  # take reciprocal when base/quote reversed
    if exp == 1:
        data["volume"] = pd.to_numeric(data["volume"])
    elif exp == -1:
        data["volume"] = pd.to_numeric(data["volume"]) / data["price"]

    # Fill in missing data with zeros
    t_samples = pd.date_range(start=t_start, end=t_end, freq="30min", tz=timezone.utc)
    data = data.reindex(t_samples, fill_value=0)

    return data


def vwap_mkt(markets, t_start, t_end):
    prices = []
    volumes = []
    for market in markets:
        data = get_mkt(market, t_start, t_end)
        prices.append(data["price"])
        volumes.append(data["volume"])

    prices = pd.concat(prices, axis=1)
    volumes = pd.concat(volumes, axis=1)

    sum_volume = volumes.sum(1)
    weights = volumes.divide(sum_volume, "index")
    vwap = (prices * weights.values).sum(1)

    data = pd.concat([vwap, sum_volume], axis=1)
    data.columns = ["price", "volume"]

    return data


def vwap_agg(coins, t_start, t_end):
    prices = []
    volumes = []

    # Using input base/quote
    data = get_agg(coins, t_start, t_end, exp=1)
    prices.append(data["price"])
    volumes.append(data["volume"])

    # Using reversed base-quote
    data = get_agg([coins[1], coins[0]], t_start, t_end, exp=-1)
    prices.append(data["price"])
    volumes.append(data["volume"] / data["price"])

    prices = pd.concat(prices, axis=1)
    volumes = pd.concat(volumes, axis=1)

    sum_volume = volumes.sum(1)
    weights = volumes.divide(sum_volume, "index")
    vwap = (prices * weights.values).sum(1)

    data = pd.concat([vwap, sum_volume], axis=1)
    data.columns = ["price", "volume"]

    return data


def update(coins, quote, t_start, t_end, pairs=False):  # noqa: C901
    t_start = t_start.replace(tzinfo=timezone.utc)
    t_start_orig = t_start
    t_end = t_end.replace(tzinfo=timezone.utc)

    # If no quote, get prices for each pair of coins
    if quote is None:
        if pairs is False:
            combos = list(combinations(coins, 2))
        else:
            combos = coins

        for pair in combos:
            print("Downloading " + pair[0] + "-" + pair[1])
            try:
                curr_file = pd.read_csv("data/" + pair[0] + "-" + pair[1] + ".csv", index_col=0)
                curr_file.index = pd.to_datetime(curr_file.index)

                if t_start_orig < curr_file.index[-1]:
                    t_start = pd.to_datetime(curr_file.index[-1]) + timedelta(minutes=30)
                    t_start = t_start.replace(tzinfo=timezone.utc)
                else:
                    t_start = t_start_orig

                if t_start < t_end:
                    data = vwap_agg(pair, t_start, t_end)
                    data = curr_file.append(data)
                    data = data[data.index >= t_start_orig]
                    data.to_csv("data/" + pair[0] + "-" + pair[1] + ".csv")
            except Exception:
                data = vwap_agg(pair, t_start_orig, t_end)
                data.to_csv("data/" + pair[0] + "-" + pair[1] + ".csv")

    # If quote given, get prices for each coin in quote currency
    else:
        for coin in coins:
            curr_coins = [coin, quote]
            try:
                curr_file = pd.read_csv("data/" + coin + "-" + quote + ".csv", index_col=0)
                curr_file.index = pd.to_datetime(curr_file.index)

                if t_start_orig < curr_file.index[-1]:
                    t_start = pd.to_datetime(curr_file.index[-1]) + timedelta(minutes=30)
                    t_start = t_start.replace(tzinfo=timezone.utc)
                else:
                    t_start = t_start_orig

                if t_start < t_end:
                    data = vwap_agg(curr_coins, t_start, t_end)
                    data = curr_file.append(data)
                    data = data[data.index >= t_start_orig]
                    data.to_csv("data/" + coin + "-" + quote + ".csv")
            except Exception:
                data = vwap_agg(curr_coins, t_start_orig, t_end)
                data.to_csv("data/" + coin + "-" + quote + ".csv")


def poolprices(  # noqa: C901
    coins=[], quote=None, quotediv=False, t_start=None, t_end=None, resample=None, pairs=[], data_dir="data"
):
    """
    Loads and formats price/volume data from CSVs.

    coins: list of coins to load (e.g., ['DAI', 'USDC', 'USDT'])
    quote: if string, name of quote currency to load (e.g., 'USD')
    quotediv: determine pairwise coin prices using third currency (e.g., ETH-SUSD/SETH-SUSD for ETH-SETH)
    t_start/t_end: used to truncate input time series
    resample: used to downsample input time series
    pairs: list of coin pairs to load (e.g., ['DAI-USDC', 'USDC-USDT'])
    data_dir: base directory name for price csv files

    Returns exchange rates/volumes for each coin pair in order of list(itertools.combinations(coins,2))

    """
    if pairs and coins:
        raise ValueError("Use only 'coins' or 'pairs', not both.")

    if coins:
        if quote:
            symbol_pairs = zip(coins, [quote] * len(coins))
        else:
            symbol_pairs = list(combinations(coins, 2))
    elif pairs:
        symbol_pairs = pairs
    else:
        raise ValueError("Must use one of 'coins' or 'pairs'.")

    prices = []
    volumes = []
    for (sym_1, sym_2) in symbol_pairs:
        filename = os.path.join(data_dir, f"{sym_1}-{sym_2}.csv")
        data_df = pd.read_csv(filename, index_col=0)
        prices.append(data_df["price"])
        volumes.append(data_df["volume"])

    prices = pd.concat(prices, axis=1)
    volumes = pd.concat(volumes, axis=1)

    pzero = (prices == 0).mean()

    prices = prices.replace(to_replace=0, method="ffill")  # replace price=0 with previous price
    prices = prices.replace(
        to_replace=0, method="bfill"
    )  # replace any price=0 at beginning with subsequent price

    # If quotediv, calc prices for each coin pair from prices in quote currency
    if quotediv:
        combos = list(combinations(range(len(coins)), 2))
        prices_tmp = []
        volumes_tmp = []

        for pair in combos:
            prices_tmp.append(prices.iloc[:, pair[0]] / prices.iloc[:, pair[1]])  # divide prices
            volumes_tmp.append(volumes.iloc[:, pair[0]] + volumes.iloc[:, pair[1]])  # sum volumes

        prices = pd.concat(prices_tmp, axis=1)
        volumes = pd.concat(volumes_tmp, axis=1)

    # Index as date-time type
    prices.index = pd.to_datetime(prices.index)
    volumes.index = pd.to_datetime(volumes.index)

    # Trim to t_start and/or t_end
    if t_start is not None:
        prices = prices.loc[t_start:]
        volumes = volumes.loc[t_start:]

    if t_end is not None:
        prices = prices.loc[:t_end]
        volumes = prices.loc[:t_end]

    # Resample times
    if resample is not None:
        prices = prices.resample(resample).first()
        volumes = volumes.resample(resample).sum()
    else:
        prices.index.freq = pd.infer_freq(prices.index)
        volumes.index.freq = pd.infer_freq(volumes.index)

    return prices, volumes, pzero
